src/shogun/base/subset_iterators.h

```diff
@@ -0,0 +1,191 @@
+			using pointer = typename std::conditional_t<
+			    is_const,
+			    const typename return_type_subset_iterator<
+			        remove_cvptr_t<T>>::type*,
+			    typename return_type_subset_iterator<remove_cvptr_t<T>>::type*>;
+
+			using internal_pointer = typename std::conditional_t<
+			    is_const, const remove_cvptr_t<T>*, remove_cvptr_t<T>*>;
+
+			subset_iterator(internal_pointer ptr, index_t idx = 0)
+			{
+				if (ptr->get_subset_stack()->get_last_subset() != nullptr)
+					m_subset = ptr->get_subset_stack()
+					               ->get_last_subset()
+					               ->get_subset_idx();
+				m_idx = idx;
+				// need to keep this variable so that the vector destructor
+				// isn't called
+				m_argsorted_subset = CMath::argsort(m_subset);
+				m_argsorted_subset_iter = m_argsorted_subset.begin() + idx;
+				m_ptr = ptr;
+			}
+
+			subset_iterator<T, is_const>& operator++()
+			{
+				m_argsorted_subset_iter++;
+				m_idx++;
+				return *this;
+			}
+
+			subset_iterator<T, is_const> operator++(int)
+			{
+				subset_iterator<T, is_const> retval(*this);
+				++(*this);
+				return retval;
+			}
+
+			bool operator==(const subset_iterator<T, is_const>& other)
+			{
+				return m_ptr == other.m_ptr && m_idx == other.m_idx;
+			}
+
+			bool operator!=(const subset_iterator<T, is_const>& other)
+			{
+				return !(*this == other);
+			}
+
+			reference operator*()
+			{
+				index_t idx;
+				if (m_ptr->get_subset_stack()->get_last_subset() != nullptr)
+					idx = *m_argsorted_subset_iter;
+				else
+					idx = m_idx;
+				if constexpr (std::is_base_of<
+				                  CFeatures, remove_cvptr_t<T>>::value)
+					return m_ptr->get_feature_vector(idx);
+				if constexpr (std::is_base_of<
+				                  CLabels, remove_cvptr_t<T>>::value)
+					return m_ptr->get_label(idx);
+			}
+
+		private:
+			index_t m_idx;
+			internal_pointer m_ptr;
+			SGVector<index_t> m_subset;
+			SGVector<index_t> m_argsorted_subset;
+			SGVector<index_t>::iterator m_argsorted_subset_iter;
+		};
+
+		/**
+		 * Returns an iterator to the first vector of features/labels.
+		 */
+		auto begin()
+		{
+			auto* this_casted = static_cast<IterableSubsetContainer<ST>*>(this);
+			return subset_iterator<decltype(this_casted)>(
+			    static_cast<IterableSubsetContainer<ST>*>(this));
+		}
+
+		/**
+		 * Returns an iterator to the element following the end of
+		 * features/labels.
+		 */
+		auto end()
+		{
+			auto* this_casted = static_cast<IterableSubsetContainer<ST>*>(this);
+			if (auto stack = this_casted->get_subset_stack()->get_last_subset();
+			    stack == nullptr)
+				return subset_iterator<decltype(this_casted)>(
+				    this_casted, this_casted->get_num_vectors());
+			else
+				return subset_iterator<decltype(this_casted)>(
+				    this_casted, stack->get_subset_idx().size());
+		}
+
+		/**
+		 * Returns a const iterator to the first vector of features/labels.
+		 */
+		auto begin() const
```

Comment:
@karlnapf the range loop calls the const version if it's a const pointer

Reply:
sweet, very nice template magic!